stdlib/float.ml

```diff
@@ -92,11 +92,361 @@ external seeded_hash_param : int -> int -> int -> float -> int
+      bubbledown l j
+    in
+    let bubble l i = try bubbledown l i with Bottom i -> i in
+    let rec trickleup i e =
+      let father = (i - 1) / 3 in
+      assert (i <> father);
+      if cmp (get a father) e < 0 then begin
+        set a i (get a father);
+        if father > 0 then trickleup father e else set a 0 e;
+      end else begin
+        set a i e;
+      end;
+    in
+    let l = length a in
+    for i = (l + 1) / 3 - 1 downto 0 do trickle l i (get a i); done;
+    for i = l - 1 downto 2 do
+      let e = (get a i) in
+      set a i (get a 0);
+      trickleup (bubble i 0) e;
+    done;
+    if l > 1 then (let e = (get a 1) in set a 1 (get a 0); set a 0 e)
+
+  (* duplicated from array.ml, except for the call to [create] *)
+  let cutoff = 5
+  let stable_sort cmp a =
+    let merge src1ofs src1len src2 src2ofs src2len dst dstofs =
+      let src1r = src1ofs + src1len and src2r = src2ofs + src2len in
+      let rec loop i1 s1 i2 s2 d =
+        if cmp s1 s2 <= 0 then begin
+          set dst d s1;
+          let i1 = i1 + 1 in
+          if i1 < src1r then
+            loop i1 (get a i1) i2 s2 (d + 1)
+          else
+            blit src2 i2 dst (d + 1) (src2r - i2)
+        end else begin
+          set dst d s2;
+          let i2 = i2 + 1 in
+          if i2 < src2r then
+            loop i1 s1 i2 (get src2 i2) (d + 1)
+          else
+            blit a i1 dst (d + 1) (src1r - i1)
+        end
+      in loop src1ofs (get a src1ofs) src2ofs (get src2 src2ofs) dstofs;
+    in
+    let isortto srcofs dst dstofs len =
+      for i = 0 to len - 1 do
+        let e = (get a (srcofs + i)) in
+        let j = ref (dstofs + i - 1) in
+        while (!j >= dstofs && cmp (get dst !j) e > 0) do
+          set dst (!j + 1) (get dst !j);
+          decr j;
+        done;
+        set dst (!j + 1) e;
+      done;
+    in
+    let rec sortto srcofs dst dstofs len =
+      if len <= cutoff then isortto srcofs dst dstofs len else begin
+        let l1 = len / 2 in
+        let l2 = len - l1 in
+        sortto (srcofs + l1) dst (dstofs + l1) l2;
+        sortto srcofs a (srcofs + l2) l1;
+        merge (srcofs + l2) l1 dst (dstofs + l1) l2 dst dstofs;
+      end;
+    in
+    let l = length a in
+    if l <= cutoff then isortto 0 a 0 l else begin
+      let l1 = l / 2 in
+      let l2 = l - l1 in
+      let t = create l2 in
+      sortto l1 t 0 l2;
+      sortto 0 a l2 l1;
+      merge l2 l1 t 0 l2 a 0;
+    end
+
+  let fast_sort = stable_sort
+
+  (* duplicated from array.ml *)
+  let to_seq a =
+    let rec aux i () =
+      if i < length a
+      then
+        let x = unsafe_get a i in
+        Seq.Cons (x, aux (i+1))
+      else Seq.Nil
+    in
+    aux 0
+
+  (* duplicated from array.ml *)
+  let to_seqi a =
+    let rec aux i () =
+      if i < length a
+      then
+        let x = unsafe_get a i in
+        Seq.Cons ((i,x), aux (i+1))
+      else Seq.Nil
+    in
+    aux 0
+
+  (* mostly duplicated from array.ml *)
```

Comment:
(The "mostly" is related to the special casing of
the empty array. Shall keep the special casing
by introducing a constant?)

Reply:
(Would also apply to `map`, `map2`, and `mapi`.)

Reply:
There's also the `create` that takes only one argument.

Reply:
Sorry, I meant "mostly" because the version from `Array`
starts with: `function [] -> [||] | ...`. So, I wondered
whether it would make sens to have a similar fast track
here, possibly by defining `let empty = create 0`.